events/janus_gelfevh.c

```diff
@@ -0,0 +1,617 @@
+ */
+
+#include "eventhandler.h"
+
+#include <math.h>
+
+#include "../debug.h"
+#include "../config.h"
+#include "../mutex.h"
+#include "../utils.h"
+#include "../events.h"
+#include <netdb.h>
+#include <errno.h>
+#include <sys/time.h>
+#include <arpa/inet.h>
+#include <sys/socket.h>
+#include <netinet/in.h>
+#include "../ip-utils.h"
+
+/* Plugin information */
+#define JANUS_GELFEVH_VERSION			1
+#define JANUS_GELFEVH_VERSION_STRING 	"0.0.1"
+#define JANUS_GELFEVH_DESCRIPTION 		"This is event handler plugin for Janus, which forwards events via TCP/UDP to GELF server."
+#define JANUS_GELFEVH_NAME 				"JANUS GelfEventHandler plugin"
+#define JANUS_GELFEVH_AUTHOR 			"Meetecho s.r.l."
+#define JANUS_GELFEVH_PACKAGE			"janus.eventhandler.gelfevh"
+
+#define MAX_GELF_CHUNKS 				128
+
+/* Plugin methods */
+janus_eventhandler *create(void);
+int janus_gelfevh_init(const char *config_path);
+void janus_gelfevh_destroy(void);
+int janus_gelfevh_get_api_compatibility(void);
+int janus_gelfevh_get_version(void);
+const char *janus_gelfevh_get_version_string(void);
+const char *janus_gelfevh_get_description(void);
+const char *janus_gelfevh_get_name(void);
+const char *janus_gelfevh_get_author(void);
+const char *janus_gelfevh_get_package(void);
+void janus_gelfevh_incoming_event(json_t *event);
+json_t *janus_gelfevh_handle_request(json_t *request);
+
+/* Event handler setup */
+static janus_eventhandler janus_gelfevh =
+	JANUS_EVENTHANDLER_INIT (
+		.init = janus_gelfevh_init,
+		.destroy = janus_gelfevh_destroy,
+
+		.get_api_compatibility = janus_gelfevh_get_api_compatibility,
+		.get_version = janus_gelfevh_get_version,
+		.get_version_string = janus_gelfevh_get_version_string,
+		.get_description = janus_gelfevh_get_description,
+		.get_name = janus_gelfevh_get_name,
+		.get_author = janus_gelfevh_get_author,
+		.get_package = janus_gelfevh_get_package,
+
+		.incoming_event = janus_gelfevh_incoming_event,
+		.handle_request = janus_gelfevh_handle_request,
+
+		.events_mask = JANUS_EVENT_TYPE_NONE
+	);
+
+/* Plugin creator */
+janus_eventhandler *create(void) {
+	JANUS_LOG(LOG_VERB, "%s created!\n", JANUS_GELFEVH_NAME);
+	return &janus_gelfevh;
+}
+
+/* Compression, if any */
+static gboolean compress = FALSE;
+static int compression = 6; /* Z_DEFAULT_COMPRESSION */
+
+/* Useful stuff */
+static volatile gint initialized = 0, stopping = 0;
+static GThread *handler_thread;
+static void *janus_gelfevh_handler(void *data);
+static janus_mutex evh_mutex;
+
+/* JSON serialization options */
+static size_t json_format = JSON_INDENT(3) | JSON_PRESERVE_ORDER;
+
+/* Queue of events to handle */
+static GAsyncQueue *events = NULL;
+static json_t exit_event;
+static void janus_gelfevh_event_free(json_t *event) {
+	if(!event || event == &exit_event)
+		return;
+	json_decref(event);
+}
+
+/* Gelf backend to send the events to */
+static char *backend = NULL;
+static char *port = NULL;
+static int max_gelf_msg_len = 500;
+static int janus_gelfevh_socket_type = 1;
```

Comment:
This should either be an `enum`, to explicitly say "this is UDP, this is TCP" (helpful in case this can be expanded later on with more protocols) or a boolean with the protocol name in it (e.g., `janus_gelfevh_use_tcp` that when `TRUE` means you'll use TCP). A generically named variable with a number value doesn't tell who reads the code what that variable does.

Reply:
Please address this too.

Reply:
ok lets re-build this to enum

Reply:
I can't seem to set it as enum, since it is already set that way, so I get this error:
```
events/janus_gelfevh.c:111:2: error: redeclaration of enumerator ‘SOCK_STREAM’
  SOCK_STREAM,
  ^
/usr/include/x86_64-linux-gnu/bits/socket_type.h:26:3: note: previous definition of ‘SOCK_STREAM’ was here
   SOCK_STREAM = 1,  /* Sequenced, reliable, connection-based
   ^
events/janus_gelfevh.c:112:2: error: redeclaration of enumerator ‘SOCK_DGRAM’
  SOCK_DGRAM
  ^
/usr/include/x86_64-linux-gnu/bits/socket_type.h:29:3: note: previous definition of ‘SOCK_DGRAM’ was here
   SOCK_DGRAM = 2,  /* Connectionless, unreliable datagrams
   ^
```
I know currently is not so nicely done, but if I put an enum as TCP/UDP then I need to do transformation to SOCK_* or numbers 1/2 to pass to connect, so one more if/else.
Not sure what is smart here.

Reply:
ok solved